sdk/identity/Azure.Identity/src/ClientCertificateCredential.cs

```diff
@@ -135,5 +183,215 @@ public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext r
+
+            public X509Certificate2FromObjectProvider(X509Certificate2 clientCertificate)
+            {
+                Certificate = clientCertificate ?? throw new ArgumentNullException(nameof(clientCertificate));
+            }
+
+            public Task<X509Certificate2> GetCertificateAsync(bool async, CancellationToken cancellationToken)
+            {
+                return Task.FromResult(Certificate);
+            }
+        }
+
+        /// <summary>
+        /// X509Certificate2FromFileProvider provides an X509Certificate2 from a file on disk.  It supports both
+        /// "pfx" and "pem" encoded certificates.
+        /// </summary>
+        internal class X509Certificate2FromFileProvider : IX509Certificate2Provider
+        {
+            // Lazy initialized on the first call to GetCertificateAsync, based on CertificatePath.
+            private X509Certificate2 Certificate { get; set; }
+            internal string CertificatePath { get; }
+
+            public X509Certificate2FromFileProvider(string clientCertificatePath)
+            {
+                CertificatePath = clientCertificatePath ?? throw new ArgumentNullException(nameof(clientCertificatePath));
+            }
+
+            public Task<X509Certificate2> GetCertificateAsync(bool async, CancellationToken cancellationToken)
+            {
+                if (!(Certificate is null))
+                {
+                    return Task.FromResult(Certificate);
+                }
+
+                string fileType = Path.GetExtension(CertificatePath);
+
+                switch (fileType.ToLowerInvariant())
+                {
+                    case ".pfx":
+                        return LoadCertificateFromPfxFileAsync(async, CertificatePath, cancellationToken);
+                    case ".pem":
+                        return LoadCertificateFromPemFileAsync(async, CertificatePath, cancellationToken);
+                    default:
+                        throw new CredentialUnavailableException("Only .pfx and .pem files are supported.");
+                }
+            }
+
+            private const int BufferSize = 4 * 1024;
+
+            private async Task<X509Certificate2> LoadCertificateFromPfxFileAsync(bool async, string clientCertificatePath, CancellationToken cancellationToken)
+            {
+                if (!(Certificate is null))
+                {
+                    return Certificate;
+                }
+
+                try
+                {
+                    if (!async)
+                    {
+                        Certificate = new X509Certificate2(clientCertificatePath);
+                    }
+                    else
+                    {
+                        List<byte> certContents = new List<byte>();
+                        byte[] buf = new byte[BufferSize];
+                        int offset = 0;
+                        using (Stream s = File.OpenRead(clientCertificatePath))
```

Comment:
File.ReadAllBytes?

Reply:
I guess only core has the async overload.

Reply:
Yeah.  Not sure if we want to promote this to some sort of shared utils code or not.

Reply:
Going to keep this as is for now - if we want to pull this out into shared code later I can do that.